src/openms/source/ANALYSIS/OPENSWATH/PeakIntegrator.cpp

```diff
@@ -375,4 +375,706 @@ namespace OpenMS
+      }
+      std::cout << std::endl;
+      std::cout << "result=" << result << std::endl;
+    }
+    return result;
+  }
+
+  double PeakIntegrator::E_wrt_tau(
+    const std::vector<double>& xs,
+    const std::vector<double>& ys,
+    const double h,
+    const double mu,
+    const double sigma,
+    const double tau
+  ) const
+  {
+    const double u = mu;
+    const double s = sigma;
+    const double t = tau;
+    std::vector<double> diffs(xs.size());
+    const double PI = OpenMS::Constants::PI;
+    for (Size i = 0; i < xs.size(); ++i)
+    {
+      const double x = xs[i];
+      const double y = ys[i];
+      const double z = compute_z(x, mu, sigma, tau);
+      if (z < 0)
+      {
+        diffs[i] = (2 * (-(std::sqrt(PI/2.0) * h * s * std::exp(std::pow(s,2.0)/(2 * std::pow(t,2.0)) - (x - u)/t) * std::erfc((s/t - (x - u)/s)/std::sqrt(2.0)))/std::pow(t,2.0) + (std::sqrt(PI/2.0) * h * s * std::exp(std::pow(s,2.0)/(2.0 * std::pow(t,2.0)) - (x - u)/t) * ((x - u)/std::pow(t,2.0) - std::pow(s,2.0)/std::pow(t,3.0)) * std::erfc((s/t - (x - u)/s)/std::sqrt(2.0)))/t + (h * std::pow(s,2.0) * std::exp(std::pow(s,2.0)/(2.0 * std::pow(t,2.0)) - 1.0/2.0 * std::pow((s/t - (x - u)/s),2.0) - (x - u)/t))/std::pow(t,3.0)) * ((std::sqrt(PI/2.0) * h * s * std::exp(std::pow(s,2.0)/(2.0 * std::pow(t,2.0)) - (x - u)/t) * std::erfc((s/t - (x - u)/s)/std::sqrt(2.0)))/t - y)) / static_cast<double>(xs.size());
+      }
+      else if (z <= 6.71e7)
+      {
+        diffs[i] = (2 * (-(std::sqrt(PI/2.0) * h * std::pow(s,2.0) * std::exp(1.0/2.0 * std::pow((s/t - (x - u)/s),2.0) - std::pow(x-u,2.0)/(2.0 * std::pow(s,2.0))) * (s/t - (x - u)/s) * std::erfc((s/t - (x - u)/s)/std::sqrt(2.0)))/std::pow(t,3.0) - (std::sqrt(PI/2.0) * h * s * std::exp(1.0/2.0 * std::pow((s/t - (x - u)/s),2.0) - std::pow(x-u,2.0)/(2.0 * std::pow(s,2.0))) * std::erfc((s/t - (x - u)/s)/std::sqrt(2.0)))/std::pow(t,2.0) + (h * std::pow(s,2.0) * std::exp(-std::pow(x-u,2.0)/(2 * std::pow(s,2.0))))/std::pow(t,3.0)) * ((std::sqrt(PI/2.0) * h * s * std::exp(1.0/2.0 * std::pow((s/t - (x - u)/s),2.0) - std::pow(x-u,2.0)/(2 * std::pow(s,2.0))) * std::erfc((s/t - (x - u)/s)/std::sqrt(2.0)))/t - y)) / static_cast<double>(xs.size());
+      }
+      else
+      {
+        diffs[i] = ((2.0 * h * (x - u) * std::exp(-std::pow(x-u,2.0)/(2.0 * std::pow(s,2.0))) * ((h * std::exp(-std::pow(x-u,2.0)/(2.0 * std::pow(s,2.0))))/(1.0 - (t * (x - u))/std::pow(s,2.0)) - y))/(std::pow(s,2.0) * std::pow((1.0 - (t * (x - u))/std::pow(s,2.0)),2.0))) / static_cast<double>(xs.size());
+      }
+    }
+    const double result = std::accumulate(diffs.begin(), diffs.end(), 0.0);
+    if (print_debug_ == 2)
+    {
+      std::cout << std::endl << "E_wrt_tau() diffs:" << std::endl;
+      for (double d : diffs)
+      {
+        std::cout << d << " ";
+      }
+      std::cout << std::endl;
+      std::cout << "result=" << result << std::endl;
+    }
+    return result;
+  }
+
+  double PeakIntegrator::Loss_function(
```

Comment:
Please do add unit tests with 3 cases for each range of z.  It would make for a horrible feeling if we found out that there was a typo in one of the equations.